instrumentation/opentelemetry_phoenix/lib/opentelemetry_phoenix.ex

```diff
@@ -0,0 +1,221 @@
+    :telemetry.attach(
+      {__MODULE__, :endpoint_start},
+      opts[:endpoint_prefix] ++ [:start],
+      &__MODULE__.handle_endpoint_start/4,
+      %{}
+    )
+  end
+
+  @doc false
+  def attach_endpoint_stop_handler(opts) do
+    :telemetry.attach(
+      {__MODULE__, :endpoint_stop},
+      opts[:endpoint_prefix] ++ [:stop],
+      &__MODULE__.handle_endpoint_stop/4,
+      %{}
+    )
+  end
+
+  @doc false
+  def attach_router_start_handler do
+    :telemetry.attach(
+      {__MODULE__, :router_dispatch_start},
+      [:phoenix, :router_dispatch, :start],
+      &__MODULE__.handle_router_dispatch_start/4,
+      %{}
+    )
+  end
+
+  @doc false
+  def attach_router_dispatch_exception_handler do
+    :telemetry.attach(
+      {__MODULE__, :router_dispatch_exception},
+      [:phoenix, :router_dispatch, :exception],
+      &__MODULE__.handle_router_dispatch_exception/4,
+      %{}
+    )
+  end
+
+  @doc false
+  def handle_endpoint_start(_event, _measurements, %{conn: %{adapter: adapter} = conn} = meta, _config) do
+    # TODO: maybe add config for what paths are traced? Via sampler?
+    :otel_propagator.text_map_extract(conn.req_headers)
+
+    peer_data = Plug.Conn.get_peer_data(conn)
+
+    user_agent = header_value(conn, "user-agent")
+    peer_ip = Map.get(peer_data, :address)
+
+    attributes = [
+      "http.client_ip": client_ip(conn),
+      "http.flavor": http_flavor(adapter),
+      "http.host": conn.host,
+      "http.method": conn.method,
+      "http.scheme": "#{conn.scheme}",
+      "http.target": conn.request_path,
+      "http.user_agent": user_agent,
+      "net.host.ip": to_string(:inet_parse.ntoa(conn.remote_ip)),
+      "net.host.port": conn.port,
+      "net.peer.ip": to_string(:inet_parse.ntoa(peer_ip)),
+      "net.peer.port": peer_data.port,
+      "net.transport": :"IP.TCP"
+    ]
+
+    # start the span with a default name. Route name isn't known until router dispatch
+    OpentelemetryTelemetry.start_telemetry_span(@tracer_id, "HTTP #{conn.method}", meta, %{kind: :server})
+    |> Span.set_attributes(attributes)
+  end
+
+  @doc false
+  def handle_endpoint_stop(_event, _measurements, %{conn: conn} = meta, _config) do
+    # ensure the correct span is current and update the status
+    ctx = OpentelemetryTelemetry.set_current_telemetry_span(@tracer_id, meta)
+
+    Span.set_attribute(ctx, :"http.status", conn.status)
+
+    if conn.status >= 400 do
+      Span.set_status(ctx, OpenTelemetry.status(:error, ""))
+    end
+
+    # end the Phoenix span
+    OpentelemetryTelemetry.end_telemetry_span(@tracer_id, meta)
+  end
+
+  @doc false
+  def handle_router_dispatch_start(_event, _measurements, meta, _config) do
+    attributes = [
+      "phoenix.plug": meta.plug,
+      "phoenix.action": meta.plug_opts,
+      "http.route": meta.route
+    ]
+
+    # Add more info that we now know about but don't close the span
+    ctx = OpentelemetryTelemetry.set_current_telemetry_span(@tracer_id, meta)
```

Comment:
unrelated to the migration: I wonder if this should be inside a try/catch block or if there's should be a mention of `Plug.Telemetry` requirement in the README to avoid this cryptic error: https://github.com/opentelemetry-beam/opentelemetry_phoenix/issues/34#issuecomment-905275488

Reply:
Feel free to add a doc PR to note the req for plug telemetry.